### ISSUE TITLE
[AST] Consider expanding the use of NatSpec documentation to all declarations.

### ISSUE THREAD

**Issue by christianparpart** [ethereum/solidity#12295]
The motivation behind is, to be able to display additional information about a declaration, not just its type signature.

This might also help documentation.

The relevant AST nodes I'd at least expand to are:

- [x] struct definitions
- [x] enum definitions
- [ ] enum value definitions
- [ ] UDVTs
- [ ] ASTJSON tests for custom errors
- [ ] modules
- [ ] `constant` variable definitions

**Comment by frangio** [comment]
Would really like to see this. On top of the ones mentioned in the issue, I'd add custom errors and user defined value types.

It's currently weird because it's possible to add NatSpec in the source code but it doesn't show up in the AST.

**Comment by chriseth** [comment]
Sounds good!

**Comment by PaulRBerg** [comment]
An interim solution is to apply a `@custom` NatSpec tag to document struct fields and enum values. Though I am not sure if I should do it like this:

```solidity
/// @custom: var Description for my var
struct Foo {
    uint256 var;
}
```

Or like this:

```solidity
/// @custom:field var Description for my var
struct Foo {
    uint256 var;
}
```

What would you recommend, @chriseth? I can't quite tell if there's any meaningful difference between the two, i.e. if one of the two variants produces a more informative ABI.

**Comment by PaulRBerg** [comment]
Any chance this could be included in Solidity v0.8.19?

Seems like an easy fix but with lopsided benefits for end users - e.g. OpenZeppelin is currently working on their V5. It would be super helpful to have NatSpec documentation in all OpenZeppelin structs!

At a minimum, could this be implemented for structs?

**Comment by cameel** [comment]
Probably not 0.8.19 because we want to release that pretty soon (our initial plan was to aim for 2 weeks after 0.8.18 and that may still be possible).

But as for including it in some release in the near future, I'll defer to @NunoFilipeSantos here. Unfortunately this is nowhere close to our current roadmap. This is not a big task so we might be able to fit it in somewhere between bigger tasks, but I can't really promise anything.

Incidentally, we have a pending PR that adds support for NatSpec on enums: #13626. It's stuck on some trivial things (needs more tests), mostly because everyone on the team is busy with other tasks so there and there was no one to take it over so far. If some contributor would pick that up, we might have at least that pretty soon.

**Comment by PaulRBerg** [comment]
I see, thanks anyway @cameel - I do appreciate that you are tied up nowadays.

**Comment by frangio** [comment]
This definitely isn't world-ending priority but I think we should consider it a bug with medium rather than low impact.

**Comment by cameel** [comment]
I'm fine with increasing the impact - this is only an estimate and the fact that it's of that much interest to libraries shows that it perhaps should be higher. I wouldn't really consider it a bug though because AFAIK it's was just designed that way and not originally meant for internal documentation. Of course it does not mean that we should not change it. It's just more of a minor feature at this point.

**Comment by nikola-matic** [comment]
@frangio, @PaulRBerg `struct`s and `enum`s are definitely going into 0.8.20 since they're merged, and @veniger is continuing work on custom errors.

**Comment by nikola-matic** [comment]
> Would really like to see this. On top of the ones mentioned in the issue, I'd add custom errors and user defined value types.
> 
> It's currently weird because it's possible to add NatSpec in the source code but it doesn't show up in the AST.

@frangio, replying a tad bit late, but anyway, are you sure that custom error NatSpec is not exported to the JSON AST?
Given the following minimal *.sol:
```solidity
/// @notice example of notice
/// @dev example of dev
error CustomError();
```
running with solc v0.8.19 I get the following output when using `--ast-compact-json --pretty-json`:
```json
{
  "absolutePath": "error.sol",
  "exportedSymbols":
  {
    "CustomError":
    [
      3
    ]
  },
  "id": 4,
  "nodeType": "SourceUnit",
  "nodes":
  [
    {
      "documentation":
      {
        "id": 1,
        "nodeType": "StructuredDocumentation",
        "src": "0:54:0",
        "text": "@notice example of notice\n @dev example of dev"
      },
      "errorSelector": "09caebf3",
      "id": 3,
      "name": "CustomError",
      "nameLocation": "60:11:0",
      "nodeType": "ErrorDefinition",
      "parameters":
      {
        "id": 2,
        "nodeType": "ParameterList",
        "parameters": [],
        "src": "71:2:0"
      },
      "src": "54:20:0"
    }
  ],
  "src": "54:21:0"
}
```

**Comment by frangio** [comment]
@nikola-matic That seems right, I misspoke. For user defined types I don't see the documentation in the AST.

**Comment by veniger** [comment]
@frangio I will start implementing this for the UDVT's today, you can expect a PR tommorow or monday

**Comment by PaulRBerg** [comment]
Am I understanding correctly that #14119 did _not_ add support for documenting struct params/ fields? Based on this [example](https://github.com/veniger/solidity/blob/07def48f4597559f8941a03b46b1c80925b08b9c/test/libsolidity/syntaxTests/natspec/docstring_struct.sol), only the `@title`, `@author`, `@notice`, and `@dev` seem to be supported.

Why not let users document struct fields with the `@param` tag? Or create a new tag `@field` to make a distinction between struct fields and function params?

**Comment by nikola-matic** [comment]
> Am I understanding correctly that #14119 did _not_ add support for documenting struct params/ fields? Based on this [example](https://github.com/veniger/solidity/blob/07def48f4597559f8941a03b46b1c80925b08b9c/test/libsolidity/syntaxTests/natspec/docstring_struct.sol), only the `@title`, `@author`, `@notice`, and `@dev` seem to be supported.
> 
> Why not let users document struct fields with the `@param` tag? Or create a new tag `@field` to make a distinction between struct fields and function params?

Yup, you are correct; at the moment, only top level documentation for `struct`s is exported. NatSpec is Doxygen inspired, which generally serves to document the interface, rather than the implementation itself, so I am personally not a big fan of having custom tags for contract/struct internals, however, I do agree that having a specific tag for such scenarios (e.g. `@field` that you suggested) is preferable than misusing `@param` for that purpose.

PRs for UDVTs and enum values are open right now and will be part of the next release; beyond that, we have no plans to work on NatSpec for the foreseeable future (as far as I'm aware). You are of course welcome in the Wednesday language design meetings to convince us otherwise :)

**Comment by frangio** [comment]
Struct fields become a part of the interface if they are used as a function argument or return value though?

**Comment by gabrielkrell** [comment]
> PRs for UDVTs and enum values are open right now and will be part of the next release; beyond that, we have no plans to work on NatSpec for the foreseeable future (as far as I'm aware).

@nikola-matic What happened to this? Next release has come and gone. Yesterday I had to implement some horrible local bodge to inject UDVT docstrings into my company's generated documentation. I would really like to see #14166 merged. The JSON output for UDVTs has no tags at all. At least we can document enum values with `@custom` while we wait for #14193.

------

Like frangio I also think struct fields should get first-class support, not the `custom` DIY stuff everyone is doing right now. Consider the following:

```solidity
struct User {
	uint assets;
	uint debt;
	uint64 age;
	...
}

mapping (bytes32 userId => User) private users;

function registerUser(User newUser) external onlyAdmin() returns (bytes32 userId);
function updateUser(bytes32 userId, User updatedUser) external onlyAdmin();

function getUser(bytes32 userId) external view returns (User selectedUser);
function getAverageUser() external view returns (User averageUser);
function getMedianUser() external view returns (User medianUser);
```

Clearly User is part of the interface and people calling these functions need to know what it is. But functions that work with shared structs shouldn't _each_ be responsible for documenting every field of the struct. You'd have tons of duplicate documentation, and each function would have to have many field docstrings crammed into a single `@param` or `@return`. Those functions should simply say "this takes/returns a User which represents blah" and then you click on User to see how it's constructed. If you are going to support enum values, the same logic should lead you to support struct fields.

**Comment by nikola-matic** [comment]
@gabrielkrell we're working on Cancun support (transient storage, mcopy, etc.) at the moment, hoping to release in Jan 2024 - so this is pretty low priority; in other words, no one on the team is going to be working on this for the next 1-2 months. Would you like to give it a shot? There are a few other PRs in this issue that address the same problem for structs, enums, etc.

**Comment by gabrielkrell** [comment]
@nikola-matic Unfortunately my team probably can't spare the bandwidth either. I'm glad to know it's still on your radar.

**Comment by cameel** [comment]
Continuing discussion from #16318:

@mudgen
> 3. Improved NatSpec documentation for modules. I currently cannot write a NatSpec comment for a constant defined in a module. I can't write a NatSpec comment at the top of the module to document the module itself.
>
>     For example I would like to write the following NatSpec comment at the top of the module to document the module:
>
>     ```solidity
>     /**
>     * @title Diamond Module
>     * @notice Internal functions and storage for diamond proxy functionality.
>     * @dev Implements EIP-2535 Diamond Standard
>     * (https://eips.ethereum.org/EIPS/eip-2535)
>     */
>     ```

I guess we can add modules and constants (file- and contract-level) to the list.

One problem I see with modules though - how to distinguish that it's the module being documented and not some contract that happens to be at the beginning of the file?

For example, does this document the contract or the module?
```solidity
/// @title My Module

contract C {}
```

Perhaps we could solve it by attaching the docs to the pragma:

```solidity
/// @title My Module
pragma solidity 0.8.31;

contract C {}
```
but the problem is that this will likely not be obvious to users. And that the pragma does not have to be the first thing in the file either, potentially making things quite confusing.

**Comment by mudgen** [comment]
Yes, would really like natspec tags for Solidity constants and modules.

**Comment by PaulRBerg** [comment]
> Yes, would really like natspec tags for Solidity constants and modules.

Yes, please!

In PRBMath, I have dedicated `Constants.sol` files. It would be great if Solidity provides NatSpec supports for constants like these:

https://github.com/PaulRBerg/prb-math/blob/130e0b93b94f8ff619a33eab3fb8a98ab09333ec/src/ud60x18/Constants.sol
